packages/pyright-internal/src/languageService/hoverProvider.ts

```diff
@@ -270,76 +278,134 @@ export class HoverProvider {
+                                typeText = '(module) ' + infoNode.d.value;
+                            } else {
+                                let label = 'function';
+                                let isProperty = false;
+
+                                if (isMaybeDescriptorInstance(type, /* requireSetter */ false)) {
+                                    isProperty = true;
+                                    label = 'property';
+                                }
+
+                                typeText = getToolTipForType(
+                                    type,
+                                    label,
+                                    infoNode.d.value,
+                                    this._evaluator,
+                                    isProperty,
+                                    this._functionSignatureDisplay
+                                );
+                            }
+
+                            this._addResultsPart(parts, typeText, /* python */ true);
+                            this._addDocumentationPart(parts, infoNode, /* resolvedDecl */ undefined);
                         }
-
-                        typeText = getToolTipForType(
-                            type,
-                            label,
-                            node.d.value,
-                            this._evaluator,
-                            isProperty,
-                            this._functionSignatureDisplay
-                        );
                     }
+                    break;
+                }
+                case ParseNodeType.String: {
+                    const type = this._evaluator.getExpectedType(infoNode)?.type;
+                    if (type !== undefined) {
+                        this._tryAddPartsForTypedDictKey(infoNode, type, parts);
+                    }
+                    break;
+                }
+                case ParseNodeType.UnaryOperation: {
+                    const typeResult = getTypeOfUnaryOperation(this._evaluator, infoNode, EvalFlags.None, undefined);
+                    this._addResultsForTypeResult(parts, typeResult);
+                    break;
+                }
+                case ParseNodeType.BinaryOperation: {
+                    const typeResult = getTypeOfBinaryOperation(this._evaluator, infoNode, EvalFlags.None, undefined);
+                    this._addResultsForTypeResult(parts, typeResult);
+                    break;
+                }
+                case ParseNodeType.AugmentedAssignment: {
+                    const typeResult = getTypeOfAugmentedAssignment(this._evaluator, infoNode, undefined);
+                    this._addResultsForTypeResult(parts, typeResult);
+                    break;
+                }
+                case ParseNodeType.Index: {
+                    const typeResult = getTypeOfIndex(this._evaluator, infoNode);
+                    this._addResultsForTypeResult(parts, typeResult);
+                    break;
+                }
```

Comment:
these cases look the same as what's in `definitionProvider`, can it be abstracted?